File: python/rexlang/builtins/math.py
import math as _math

from ..values import (
    VBuiltin,
    VFloat,
    VInt,
    Error,
    _as_float,
    value_to_string,
)


def _builtin_abs(v):
    if isinstance(v, VInt):
        return VInt(abs(v.value))
    if isinstance(v, VFloat):
        return VFloat(abs(v.value))
    raise Error(f"abs: expected number, got {value_to_string(v)}")


def _builtin_min(x):
    def inner(y):
        if isinstance(x, VInt) and isinstance(y, VInt):
            return x if x.value <= y.value else y
        if isinstance(x, VFloat) and isinstance(y, VFloat):
            return x if x.value <= y.value else y
        raise Error("min: expected matching numeric types")

    return VBuiltin("min$1", inner)


def _builtin_max(x):
    def inner(y):
        if isinstance(x, VInt) and isinstance(y, VInt):
            return x if x.value >= y.value else y
        if isinstance(x, VFloat) and isinstance(y, VFloat):
            return x if x.value >= y.value else y
        raise Error("max: expected matching numeric types")

    return VBuiltin("max$1", inner)


def _builtin_pow(x):
    if not isinstance(x, (VInt, VFloat)):
        raise Error("pow: expected number")

    def inner(y):
        if not isinstance(y, (VInt, VFloat)):
            raise Error("pow: expected number")
        return VFloat(_math.pow(float(x.value), float(y.value)))

    return VBuiltin("pow$1", inner)


def _builtin_atan2(y):
    def inner(x):
        return VFloat(_math.atan2(_as_float(y), _as_float(x)))

    return VBuiltin("atan2$1", inner)


def builtins() -> dict:
    return {
        "abs": VBuiltin("abs", _builtin_abs),
        "min": VBuiltin("min", _builtin_min),
        "max": VBuiltin("max", _builtin_max),
        "pow": VBuiltin("pow", _builtin_pow),
        "sqrt": VBuiltin("sqrt", lambda v: VFloat(_math.sqrt(_as_float(v)))),
        "sin": VBuiltin("sin", lambda v: VFloat(_math.sin(_as_float(v)))),
        "cos": VBuiltin("cos", lambda v: VFloat(_math.cos(_as_float(v)))),
        "tan": VBuiltin("tan", lambda v: VFloat(_math.tan(_as_float(v)))),
        "asin": VBuiltin("asin", lambda v: VFloat(_math.asin(_as_float(v)))),
        "acos": VBuiltin("acos", lambda v: VFloat(_math.acos(_as_float(v)))),
        "atan": VBuiltin("atan", lambda v: VFloat(_math.atan(_as_float(v)))),
        "atan2": VBuiltin("atan2", _builtin_atan2),
        "log": VBuiltin("log", lambda v: VFloat(_math.log(_as_float(v)))),
        "exp": VBuiltin("exp", lambda v: VFloat(_math.exp(_as_float(v)))),
        "pi": VFloat(_math.pi),
        "e": VFloat(_math.e),
    }

File: internal/stdlib/rexfiles/math.go
//go:build ignore

package main

import "math"

func Std_Math_toFloat(i int64) float64  { return float64(i) }
func Std_Math_round(f float64) int64    { return int64(math.Round(f)) }
func Std_Math_floor(f float64) int64    { return int64(math.Floor(f)) }
func Std_Math_ceiling(f float64) int64  { return int64(math.Ceil(f)) }
func Std_Math_truncate(f float64) int64 { return int64(f) }
func Std_Math_sqrt(f float64) float64   { return math.Sqrt(f) }

func Std_Math_abs(v any) any {
	switch n := v.(type) {
	case int64:
		if n < 0 {
			return -n
		}
		return n
	case float64:
		return math.Abs(n)
	}
	panic("abs: expected number")
}

func Std_Math_min(x, y any) any {
	switch a := x.(type) {
	case int64:
		if b, ok := y.(int64); ok {
			if a <= b {
				return a
			}
			return b
		}
	case float64:
		if b, ok := y.(float64); ok {
			if a <= b {
				return a
			}
			return b
		}
	}
	panic("min: expected matching numeric types")
}

func Std_Math_max(x, y any) any {
	switch a := x.(type) {
	case int64:
		if b, ok := y.(int64); ok {
			if a >= b {
				return a
			}
			return b
		}
	case float64:
		if b, ok := y.(float64); ok {
			if a >= b {
				return a
			}
			return b
		}
	}
	panic("max: expected matching numeric types")
}

func Std_Math_pow(x, y any) float64 {
	var xf, yf float64
	switch v := x.(type) {
	case int64:
		xf = float64(v)
	case float64:
		xf = v
	}
	switch v := y.(type) {
	case int64:
		yf = float64(v)
	case float64:
		yf = v
	}
	return math.Pow(xf, yf)
}

func Std_Math_sin(f float64) float64  { return math.Sin(f) }
func Std_Math_cos(f float64) float64  { return math.Cos(f) }
func Std_Math_tan(f float64) float64  { return math.Tan(f) }
func Std_Math_asin(f float64) float64 { return math.Asin(f) }
func Std_Math_acos(f float64) float64 { return math.Acos(f) }
func Std_Math_atan(f float64) float64 { return math.Atan(f) }
func Std_Math_atan2(y, x float64) float64 { return math.Atan2(y, x) }
func Std_Math_log(f float64) float64  { return math.Log(f) }
func Std_Math_exp(f float64) float64  { return math.Exp(f) }

var Std_Math_pi = math.Pi
var Std_Math_e = math.E
